src/services/organizeImports.ts

```diff
@@ -0,0 +1,291 @@
+/* @internal */
+namespace ts.OrganizeImports {
+    export function organizeImports(
+        sourceFile: SourceFile,
+        formatContext: formatting.FormatContext,
+        host: LanguageServiceHost,
+        cancellationToken: CancellationToken) {
+
+        // All of the (old) ImportDeclarations in the file, in syntactic order.
+        const oldImportDecls: ImportDeclaration[] = [];
+
+        forEachChild(sourceFile, node => {
+            cancellationToken.throwIfCancellationRequested();
+            if (isImportDeclaration(node)) {
+                oldImportDecls.push(node);
+            }
+            // TODO (https://github.com/Microsoft/TypeScript/issues/10020): sort *within* ambient modules (find using isAmbientModule)
+        });
+
+        if (oldImportDecls.length === 0) {
+            return [];
+        }
+
+        const usedImportDecls = removeUnusedImports(oldImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+        const sortedImportDecls = sortImports(usedImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+        const coalescedImportDecls = coalesceImports(sortedImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+
+        // All of the (new) ImportDeclarations in the file, in sorted order.
+        const newImportDecls = coalescedImportDecls;
+
+        const changeTracker = textChanges.ChangeTracker.fromContext({ host, formatContext });
+
+        // NB: Stopping before i === 0
+        for (let i = oldImportDecls.length - 1; i > 0; i--) {
+            changeTracker.deleteNode(sourceFile, oldImportDecls[i]);
+        }
+
+        if (newImportDecls.length === 0) {
+            changeTracker.deleteNode(sourceFile, oldImportDecls[0]);
+        }
+        else {
+            // Delete the surrounding trivia because it will have been retained in newImportDecls.
+            const replaceOptions = {
+                useNonAdjustedStartPosition: false,
+                useNonAdjustedEndPosition: false,
+                suffix: getNewLineOrDefaultFromHost(host, formatContext.options),
+             };
+            changeTracker.replaceNodeWithNodes(sourceFile, oldImportDecls[0], newImportDecls, replaceOptions);
+        }
+
+        const changes = changeTracker.getChanges();
+        return changes;
+    }
+
+    function removeUnusedImports(oldImports: ReadonlyArray<ImportDeclaration>) {
+        return oldImports; // TODO (https://github.com/Microsoft/TypeScript/issues/10020)
+    }
+
+    /* @internal */ // Internal for testing
+    export function sortImports(oldImports: ReadonlyArray<ImportDeclaration>) {
+        if (oldImports.length < 2) {

```

Comment:
* In `core.ts`:
```ts
export function compareBooleans(a: boolean, b: boolean): Comparison {
    return compareValues(a ? 1 : 0, b ? 1 : 0);
}
```
* Perform `length < 2` optimization inside `stableSort`
* Here:
```ts
return stableSort(oldImports, ({ moduleSpecifier: m1 }, { moduleSpecifier: m2 }) => {
    const name1 = getExternalModuleName(m1) || m1.getText();
    const name2 = getExternalModuleName(m2) || m2.getText();
    return compareBooleans(isExternalModuleNameRelative(name1), isExternalModuleNameRelative(name2)) || compareStringsCaseSensitive(name1, name2);
});
```
This would break the `non-relative vs invalid` and `relative vs invalid` tests, but I don't think it really matters how we sort invalid expressions in import positions.

Reply:
I'm not sure I understand your concerns about the current implementation.

Reply:
It's not that I spotted any bugs, just noticed that it could be done in 4 lines.

Reply:
It hadn't occurred to me to combine `Comparison`s with `||` - applying that should clean things up a bit.

One of the reasons I pulled out the struct was so that the same work wouldn't be repeated for every comparison in which a given import participated.  If that's not interesting, it can obviously be simplified.

Reply:
If you filter out invalid imports at the top, `getExternalModuleName` is just a property lookup. `isExternalModuleNameRelative` isn't very complicated either -- some simple regex and `indexOf` tests.